### PR TITLE
Refactor to work on Rocky9 and Ubuntu 22.04

### PR DESCRIPTION
**Pull request by pwalczysko**
cc @jburel

### CODE REVIEW

**Comment by pwalczysko** [comment]
Atm both Rocky9 and U22.04 are failing because of

```
   RUNNING HANDLER [ansible-role-omero-server : omero-server restart omero-server] ***
  fatal: [omero-server-rockylinux9]: FAILED! => {"changed": false, "msg": "Unable to start service omero-server: Job for omero-server.service failed because the control process exited with error code.\nSee \"systemctl status omero-server.service\" and \"journalctl -xeu omero-server.service\" for details.\n"}
```

I think that this is because of the unfinished dependency of the ``ome.ice`` dependency (?) 

cc @khaledk2 @sbesson

**Comment by sbesson** [comment]
> I think that this is because of the unfinished dependency of the ome.ice dependency (?)

That's very possible but the error message is highly unspecific about the underlying error. To appreciate it fully, you would need to run `molecule converge` against  a local Docker container and shell into this container to look at the output of the commands.

**Comment by pwalczysko** [comment]
```
 Ubuntu 22.04:
  
 CRITICAL Idempotence test failed because of the following tasks:
  *  => deploy_archive : deploy archive | unarchive
  *  => deploy_archive : deploy archive | unarchive
```
It could possibly be that the ^^^ failure of the idempotence on Ubuntu 22.04 is caused by me adding https://github.com/ome/ansible-role-omero-server/pull/69/commits/761f3e03bad0d426e05fee42eb230261c0dc577d - but, strangely, the idempotence passes on Rocky9 test. Any ideas about this @sbesson please ? I can see some hints in https://github.com/ome/ansible-role-omero-server/blob/cb5545b797bc8cef8428b1697d38a0b8b1bc23fc/molecule/ubuntu2004/molecule.yml#L26 but not sure what does that mean ?

**Comment by sbesson** [comment]
> Any ideas about this @sbesson please ?

Nothing immediately, it comes a bit as a surprise that the idempotence would fail for one platform but not the other.
Part of the issue is that you are integrating N development branches and there are really a lot of moving targets to make an informed guess on the source of the issue. Is there a way to enable more logging to highlight why the task is being marked as changed?

**Comment by pwalczysko** [comment]
The source of the failure of the idempotence seems to be the variable ``deploy_archive_filename`` from the deploy_archive role. It seems that the Ubuntu-22.04 test is attempting to unarchive the ...rhel9... tar.gz. But this is not something which succeeds imho, otherwise the ``converge`` test would fail, which it does not - the server is restarting happily on Ubuntu-22.04 according to the logs.

The problem is the value of that variable at the moment of the deploy archive task - see the logs below when a debug bit is added. Maybe the variable is not getting flushed - it seems like the ubuntu-22.04 test is alternately deploying the ...rhel9...tar.gz and ...ubuntu...tar.gz. 

This must be something to do with how these vars are used from the particular deploy archive and ice roles ? But unfortunately, do not see through this, any ideas please @khaledk2 @sbesson ?


```
2023-10-25T17:58:00.7657150Z TASK [deploy_archive : deploy archive | unarchive] *****************************
2023-10-25T17:58:08.5381043Z changed: [omero-server-rockylinux9]
2023-10-25T17:58:08.5477544Z 
2023-10-25T17:58:08.5478721Z TASK [deploy_archive : debug | unarchive] **************************************
2023-10-25T17:58:08.5928837Z ok: [omero-server-rockylinux9] => {
2023-10-25T17:58:08.5930284Z     "msg": "Orig path /opt/Ice-3.6.5-rhel9-x86_64.tar.gz dest path 172.17.0.1"
2023-10-25T17:58:08.5931339Z }
```

**Comment by pwalczysko** [comment]
Sorry, edited ^^^ https://github.com/ome/ansible-role-omero-server/pull/69#issuecomment-1779864974 as the output of the log was confusing - now it really shows the ubuntu2204 weirdness

**Comment by sbesson** [comment]
After some research, I believe you are hitting the issue described in the Note of https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_reuse_roles.html#running-role-dependencies-multiple-times-in-one-play i.e. `Using vars: results in those variables being scoped at the play level`.

So [these variables](https://github.com/khaledk2/ansible-role-ice/blob/f4c8d6f658bd98801ca6f69f4c8b59c4329fe875/meta/main.yml#L48-L56) are scoped at the play level and are consumed in the initial invocation of the `ome.deploy_archive` role caused by the addition to the role dependencies. This possibly points out to a refactoring of the `ansible-role-ice` to move away from this logic. 

Re https://github.com/ome/ansible-role-omero-server/commit/761f3e03bad0d426e05fee42eb230261c0dc577d, is it not possible to add the deploy_archive to the `requirements.yml` only to ensure you are using the branch of interest? It is possible you are starting to reach the limit of testing nested dependencies.

As a final point, the CI builds really need to be reactivated for each PR across all these roles to simplify the reviewing.

**Comment by khaledk2** [comment]
I have figured out why it failed in `Ubuntu 22.04`. It called the `deploy_archive` role two times, the first one with the `REHL` ice package regardless of the platform (`Ubuntu` and `RockyLinux`).  The second call runs with the correct ice package for each platform. 

In this scenario, `Rockylinux` passed because it called the role twice with the same package (`REHL`) so it will change at the first call only, but for `Ubuntu`, it will be called two times with two different args so it will change in the second call also.

I have deleted the role inside the  [meta/main.yml](https://github.com/khaledk2/ansible-role-omero-server/blob/01416c572d58fce906d6aadec9d035895e58a40f/meta/main.yml#L21C1-L21C27) file which I think calls the `deploy_archive ` role with `REHL` ice only and this fixes the issue.
I have tested it in my fork and it is green now for both `Ubuntu` and `Rockylinux` after applying these changes.

https://github.com/khaledk2/ansible-role-omero-server/actions/runs/6671567723

**Comment by pwalczysko** [comment]
> I have figured out why it failed in `Ubuntu 22.04`. It called the `deploy_archive` role two times, the first one with the `REHL` ice package regardless of the platform (`Ubuntu` and `RockyLinux`). The second call runs with the correct ice package for each platform.
> 
> In this scenario, `Rockylinux` passed because it called the role twice with the same correct package (`REHL`) so it will change at the first call only, but for `Ubuntu`, it will be called two times with two different args so it will change in the second call also.
> 
> I have commented out the role inside the [meta/main.yml](https://github.com/khaledk2/ansible-role-omero-server/blob/01416c572d58fce906d6aadec9d035895e58a40f/meta/main.yml#L21C1-L21C27) file which I think calls the `deploy_archive ` role with `REHL` ice only and this fixes the issue. I have tested it in my fork and it is green now for both `Ubuntu` and `Rockylinux` after applying these changes.
> 
> https://github.com/khaledk2/ansible-role-omero-server/actions/runs/6671567723

Amazing, thank you @khaledk2 - I have cherry-picked https://github.com/ome/ansible-role-omero-server/pull/69/commits/94c43cddb102d25eab6b987770827d7664d36387 and it makes the U22.04 test green here too !

thanks @sbesson and @khaledk2 for the help

**Comment by pwalczysko** [comment]
I have run the playbook with the changes here also against the RHEL 9 testing system (although not on a clean image). The test passed.

**Review by sbesson — CHANGES_REQUESTED** [review]
Overall looks good. One suggestion about the service file and the PSQL versions.

Interesting to see that the role changes here are minimal and most of the required changes are in the role dependencies. As for all other RHEL9/Ubuntu22.04 PR, this also means we'll need a chain of releases and a last commit dropping links to the branches before giving this a final review

File: templates/systemd-system-omero-server-service.j2 [hunk 1]
```diff
@@ -7,7 +7,7 @@ After=postgresql-9.4.service
 After=postgresql-9.5.service
 After=postgresql-9.6.service
 After=postgresql-10.service
-After=postgresql-11.service
+After=postgresql-13.service
```
[the remark this answers is not in the document]
Comment (pwalczysko) [reply templates/systemd-system-omero-server-service.j2:8]: > Possibly worth starting listing all intermediate versions of PostgreSQL (and remove the outdated ones) as we don't know which PSQL service will be installed?

Thank you. I took my lead from https://github.com/ome/ansible-role-postgresql/pull/34/files#diff-663391f3716a8f003f6b83f6478f08fef0ad6e9266b29704b1ae547e70736f2cR59 and produced a new list in https://github.com/ome/ansible-role-omero-server/pull/69/commits/2ff3ef24e56606c859cf49859d9c8a11280b7705

[the remark this answers is not in the document]
Comment (sbesson) [reply templates/systemd-system-omero-server-service.j2:8]: :+1: I think this matches https://endoflife.date/postgresql. Arguably, you could add PSQL 15 to this list which has now been released 1 year ago

**Review by sbesson — APPROVED** [review]

**Comment by pwalczysko** [comment]
The last two commits should be pointing this role to ome namespace in galaxy for all dependency roles.

But I am failing on a dependency variable declaration

```
   
  TASK [ansible-role-omero-server : omero server | create system user] ***********
  fatal: [omero-server-ubuntu2204]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: {{ omero_common_basedir }}/server: 'omero_common_basedir' is undefined\n\nThe error appears to be in '/home/runner/work/ansible-role-omero-server/ansible-role-omero-server/tasks/omero-user.yml': line 4, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: omero server | create system user\n  ^ here\n"}
```

**Comment by pwalczysko** [comment]
The test is now green here. Also I was able to publish in galaxy as https://galaxy.ansible.com/ui/standalone/roles/pwalczysko/omero_server/   (pwalczysko.omero_server).

I will test tomorrow this role in playbook.

**Comment by khaledk2** [comment]
> The last two commits should be pointing this role to ome namespace in galaxy for all dependency roles.
> 
> But I am failing on a dependency variable declaration
> 
> ```
>    
>   TASK [ansible-role-omero-server : omero server | create system user] ***********
>   fatal: [omero-server-ubuntu2204]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: {{ omero_common_basedir }}/server: 'omero_common_basedir' is undefined\n\nThe error appears to be in '/home/runner/work/ansible-role-omero-server/ansible-role-omero-server/tasks/omero-user.yml': line 4, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: omero server | create system user\n  ^ here\n"}
> ```

I have a similar issue in the `omero-web` role, I fixed it by using `import_role` for the `ome.omero_common` role instead of `include_role `inside `pre_tasks.yml`.

I think you should use this approach rather than hardcode the variable.

**Comment by pwalczysko** [comment]
Tested this role in playbook, see https://github.com/ome/ansible-role-omero-web/pull/48#issuecomment-1793487604 - works as expected.

File: templates/systemd-system-omero-server-service.j2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (jburel) [reply templates/systemd-system-omero-server-service.j2:8]: psql 11 is ending today 09/11 so we should probably remove it from the list

File: molecule/rockylinux9/molecule.yml [hunk 2]
```diff
@@ -10,11 +10,16 @@ lint: |
     ansible-lint
     flake8
 platforms:
-  - name: omero-server-centos7
-    image: centos/systemd
+  - name: omero-server-rockylinux9
+    image: eniocarboni/docker-rockylinux-systemd:9
+    tmpfs:
+      - /tmp
```
[the remark this answers is not in the document]
Comment (pwalczysko) [reply molecule/rockylinux9/molecule.yml]: fixed in https://github.com/ome/ansible-role-omero-server/pull/69/commits/570833b710f7781479f311b21dea9bd829e3484c

**Comment by pwalczysko** [comment]
The comments of @khaledk2 were adressed. The python bumped to 3.9. The build is green and there are no errors&warnings on GA.

cc @jburel @khaledk2

**Comment by pwalczysko** [comment]
Do we want to merge&release this in order to enable the deployment of the server please ?

File: templates/systemd-system-omero-server-service.j2 [hunk 3]
```diff
@@ -3,11 +3,11 @@ Description=OMERO.server
...
 # dependency is also started (PostgreSQL may be on a different server)
-After=postgresql-9.4.service
-After=postgresql-9.5.service
-After=postgresql-9.6.service
-After=postgresql-10.service
 After=postgresql-11.service
+After=postgresql-12.service
+After=postgresql-13.service
+After=postgresql-14.service
+After=postgresql-15.service
```
[the remark this answers is not in the document]
Comment (pwalczysko) [reply templates/systemd-system-omero-server-service.j2]: fixed in https://github.com/ome/ansible-role-omero-server/pull/69/commits/9c844c58f0c6224782359ff2c394fa4cbdf18f65

**Pull request merged**